### PR TITLE
[Merged by Bors] - ET-3564 date filter elastic search

### PR DESCRIPTION
**Pull request by rustonaut**
- create a `mapping.json` file with the mapping we want to have
  - and update docker to use it
  - :pushpin: we also somehow need to update the deployment to somehow get that 
- add `properties.publication_date`
- add `/users/{user_id}/personalized_documents?published_after=<datetime>`
- filter when querying personalized documents to enforce `published_after <= date_published <= now`
- **not** implement the filter for memory.rs as it's not used there

### CODE REVIEW

File: web-api/src/storage/elastic.rs [hunk 1]
```diff
@@ -314,11 +320,16 @@ impl storage::Document for Storage {
                 "num_candidates": params.num_candidates,
                 "filter": {
                     "bool": {
+                        "must": {
+                            "range": {
+                                "publication_date": range,
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:325]: :art: the macro allows for trailing commas, but it is invalid json syntax and we should avoid it
```suggestion
                                "publication_date": range
```

File: web-api/src/storage/elastic.rs [hunk 2]
```diff
@@ -314,11 +320,16 @@ impl storage::Document for Storage {
...
+                            }
+                        },
                         "must_not": {
                             "ids": {
                                 "values": params.excluded.iter().map(AsRef::as_ref).collect_vec()
                             }
-                        }
+                        },
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:332]: :art:
```suggestion
                        }
```

File: web-api/src/mind.rs [hunk 3]
```diff
@@ -235,6 +235,10 @@ impl DocumentProvider {
         self.documents.get(id)
     }
 
+    fn to_documents(&self) -> Vec<Document> {
```
Comment (janpetschexain) [inline web-api/src/mind.rs:238]: ❓ it looks like most of the changes except for the `PersonalizeBy::KnnSearch` leaked into the pr? a rebase should fix that.

File: web-api/openapi/front_office.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 openapi: 3.0.3
 info:
   title: Front Office API
-  version: 1.0.0-rc6
+  version: 1.0.0-rc7
```
Comment (acrrd) [inline web-api/openapi/front_office.yaml:4]: :wrench: `/users/.../personalized_documents` spec must be updated with the new parameter.

File: web-api/openapi/back_office.yaml [hunk 5]
```diff
@@ -47,11 +47,18 @@ paths:
...
+      description: |
+        Add documents to the system.
+
+        The system will create a representation of the document that will be used to match it against the preferences of a user.
+
+        **Important note:** Currently we allow up to a maximum of 100 documents per batch size. If you need to add more, then
+        please split up the total amount of documents in separate calls, where each call contains at maximum 100 documents.
+
+        **Important note:** Documents which have no `publication_date` (in `properties`) or a `publication_date` in the future are
+        treated as unpublished and as such are filtered out from personalized documents. This allows ingesting documents which e.g.
```
Comment (acrrd) [inline web-api/openapi/back_office.yaml:59]: I don't think we want to filter out the documents that don't have a published date if `published_after` is not set.

**Comment by acrrd** [comment]
@Robert-Steiner we need a way to update the deployment with the new schema specification for elastic search

File: web-api/src/storage/elastic.rs [hunk 6]
```diff
@@ -396,6 +396,13 @@ impl storage::Document for Storage {
         params: KnnSearchParams<'a>,
     ) -> Result<Vec<models::PersonalizedDocument>, Error> {
         // https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn
+        let mut range = serde_json::Map::default();
+        range.insert("lte".into(), "now/d".into());
+        if let Some(published_after) = params.published_after {
+            range.insert("gte".into(), published_after.to_rfc3339().into());
+        }
+
+        // https://www.elastic.co/guide/en/elasticsearch/reference/8.4/knn-search.html#approximate-knn
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:405]: :art:
```suggestion
        // https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn
```

**Review by janpetschexain — APPROVED** [review]

File: web-api/openapi/back_office.yaml (hunk 5 above)
Comment (rustonaut) [reply web-api/openapi/back_office.yaml]: From slack:

-  with `published_after`  filter `after <= publication <= now`  (like currently, this excludes `publication == null`)
-  without filter `after == null || after <= now`

File: web-api/src/storage/elastic.rs [hunk 7]
```diff
@@ -396,22 +396,51 @@ impl storage::Document for Storage {
...
+            // published_after == null || published_after <= now
+            json!({
+                "bool": {
+                    "must_not": [
+                        { "ids": excluded_ids },
+                        { "range": {
+                            "properties.publication_date": {
+                                "gt": "now"
+                            }
+                        }}
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:430]: :art: we should use the same indention as in the other json blocks for readability
```suggestion
                        {
                            "ids": excluded_ids
                        },
                        {
                            "range": {
                                "properties.publication_date": {
                                    "gt": "now"
                                }
                            }
                        }
```

File: web-api/src/storage/elastic.rs [hunk 8]
```diff
@@ -396,22 +396,51 @@ impl storage::Document for Storage {
...
-                "filter": {
-                    "bool": {
-                        "must_not": {
-                            "ids": {
-                                "values": params.excluded.iter().map(AsRef::as_ref).collect_vec()
-                            }
-                        }
-                    }
-                }
+                "filter": filter,
```
Comment (janpetschexain) [inline web-api/src/storage/elastic.rs:443]: :art: no trailing commas in json
```suggestion
                "filter": filter
```

**Review by janpetschexain — APPROVED** [review]
lgtm, just some json style

**Comment by rustonaut** [comment]
bors merge

**Comment by rustonaut** [comment]
bors merge

**Comment by acrrd** [comment]
bors merge

**Pull request closed without merging**